src/pages/workspaces/workspace/workflows/WorkflowView.js

```diff
@@ -705,33 +713,76 @@ const WorkflowView = _.flow(
               div([
                 div({ style: { height: '2rem', fontWeight: 'bold' } }, ['Step 1']),
                 label(['Select root entity type:']),
-                h(Select, {
+                h(GroupedSelect, {
                   'aria-label': 'Entity type selector',
                   isClearable: false,
                   isDisabled: currentSnapRedacted || this.isSingle() || !!Utils.editWorkspaceError(ws),
                   isSearchable: true,
                   placeholder: 'Select data type...',
                   styles: { container: old => ({ ...old, display: 'inline-block', width: 200, marginLeft: '0.5rem' }) },
                   value: selectedEntityType,
-                  onChange: selection => {
-                    const value = this.updateEntityType(selection)
-                    this.setState({ entitySelectionModel: this.resetSelectionModel(value) })
+                  onChange: async ({ value, source }) => {
+                    if (source === 'snapshot') {
+                      const selectedSnapshotCachedMetadata = _.find({ snapshotName: value }, cachedSnapshotEntityMetadata)
+
+                      const selectedSnapshotEntityMetadata = selectedSnapshotCachedMetadata ?
+                        selectedSnapshotCachedMetadata.metadata :
+                        await Ajax(signal).Workspaces.workspace(namespace, workspaceName).snapshotEntityMetadata(namespace, value)
+
+                      this.setState(_.set(['modifiedConfig', 'dataReferenceName'], value))
+                      this.setState({
+                        selectedSnapshotEntityMetadata, selectedEntityType: value, selectedTableName: undefined, selectedSnapshotTableNames: _.keys(selectedSnapshotEntityMetadata),
+                        entitySelectionModel: this.resetSelectionModel(value, {}, {}, source), selectedEntitySource: source,
+                        cachedSnapshotEntityMetadata: _.concat(cachedSnapshotEntityMetadata, { snapshotName: value, metadata: selectedSnapshotEntityMetadata })
```

Comment:
This will add unconditionally add a new cache entry every time the selection changes, causing it to grow indefinitely. This could be confusing to debug, and would cause memory leaks in certain usage patterns.